### PR TITLE
EnumSerializer

### PR DESCRIPTION
**Pull request by war-in**
Closes #

## Introduced changes

- Contract can now serialize enums!

##

- [ ] This PR contains breaking changes

### CODE REVIEW

File: starknet_py/serialization/data_serializers/enum_serializer.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    {"b": 100} => [1, 100]
+    """
+
+    serializers: IndexedOrderedDict[str, CairoDataSerializer]
+
+    def deserialize_with_context(
+        self, context: DeserializationContext
+    ) -> TupleDataclass:
+        [variant_index] = context.reader.read(1)
+        variant_name, serializer = self.serializers.items()[variant_index]
```
Comment (drknzz) [inline starknet_py/serialization/data_serializers/enum_serializer.py:34]: I'm not too sure about using the new dependency. We could just iterate over the items and get the variant index.

File: starknet_py/serialization/data_serializers/enum_serializer.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    CairoDataSerializer,
+)
+from starknet_py.serialization.tuple_dataclass import TupleDataclass
+
+
+@dataclass
+class EnumSerializer(CairoDataSerializer[Dict, TupleDataclass]):
+    """
+    Serializer of enums.
+    Can serialize a dictionary.
```
Comment (drknzz) [inline starknet_py/serialization/data_serializers/enum_serializer.py:20]: What about a named tuple and a tuple dataclass? Could the serialization of them be implemented?

File: starknet_py/serialization/data_serializers/enum_serializer.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+        }
+
+        return TupleDataclass.from_dict(result_dict)
+
+    def serialize_with_context(
+        self, context: SerializationContext, value: Dict
+    ) -> Generator[int, None, None]:
+        items = list(value.items())
+        if len(items) != 1:
+            raise ValueError("Can't serialize more than one variant.")
```
Comment (drknzz) [inline starknet_py/serialization/data_serializers/enum_serializer.py:48]: `len(items) != 1` doesn't necessarily mean there are **more** than one variants, is the non-emptiness checked somewhere else?

**Review by drknzz — COMMENTED** [review]
🚀

File: starknet_py/serialization/data_serializers/enum_serializer.py (hunk 2 above)
Comment (war-in) [reply starknet_py/serialization/data_serializers/enum_serializer.py:20]: Do we really want it? Structures are serialized from dicts only

File: starknet_py/serialization/data_serializers/cairo_data_serializer.py [hunk 4]
```diff
@@ -63,3 +65,7 @@ def serialize_with_context(
         :param value: python value to serialize.
         :return: defined SerializationType.
         """
+
+    @staticmethod
+    def remove_units_from_serialized_data(serialized_data: List) -> List:
+        return list(filter(lambda x: x is not None, serialized_data))
```
Comment (drknzz) [inline starknet_py/serialization/data_serializers/cairo_data_serializer.py:71]:
```suggestion
        return [x for x in serialized_data if x is not None]
```
shorter + easier to read 😃

File: starknet_py/serialization/data_serializers/enum_serializer.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+class EnumSerializer(CairoDataSerializer[Union[Dict, TupleDataclass], TupleDataclass]):
+    """
+    Serializer of enums.
+    Can serialize a dictionary and TupleDataclass.
+    Deserializes data to a TupleDataclass.
+
+    Example:
+        {"a": 1} => [0, 1]
+        {"b": 100} => [1, 100]
+        TupleDataclass(name='a', value=100) => [0, 100]
```
Comment (drknzz) [inline starknet_py/serialization/data_serializers/enum_serializer.py:24]: I'd keep the same naming as in deserialization, i.e. `variant`

File: starknet_py/serialization/data_serializers/unit_serializer.py [hunk 6]
```diff
@@ -25,5 +25,5 @@ def deserialize_with_context(self, context: DeserializationContext) -> None:
 
     def serialize_with_context(
         self, context: SerializationContext, value: Union[int, Uint256Dict]
-    ) -> Generator[int, None, None]:  # pyright: ignore
-        raise ValueError("Unit can't be serialized.")
+    ) -> Generator[None, None, None]:  # pyright: ignore
+        yield None
```
Comment (drknzz) [inline starknet_py/serialization/data_serializers/unit_serializer.py:29]: I really like this part

File: starknet_py/serialization/data_serializers/enum_serializer.py [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+class EnumSerializer(CairoDataSerializer[Union[Dict, TupleDataclass], TupleDataclass]):
+    """
+    Serializer of enums.
+    Can serialize a dictionary and TupleDataclass.
+    Deserializes data to a TupleDataclass.
+
+    Example:
+        {"a": 1} => [0, 1]
+        {"b": 100} => [1, 100]
+        TupleDataclass(variant='a', value=100) => [0, 100]
```
Comment (drknzz) [inline starknet_py/serialization/data_serializers/enum_serializer.py:24]: Probably adding an example Enum would make sense here. Without it, it's not exactly clear, why `"a"` serializes to 0 and `"b"` serializes to 1.

**Review by drknzz — APPROVED** [review]
👍

**Pull request merged**